src/mvcategorical.jl

```diff
@@ -77,7 +77,7 @@ function Base.rand(rng::AbstractRNG, d::MvDiscreteNonParametric{T,P}) where {T,P
     p = probs(d)
     n, k = size(p)
     map(1:k) do j
-        draw = rand(rng, P)
+        draw = rand(rng, (P === Real ? Float64 : P))
```

Comment:
That seems like a quite crude heuristic. Ideally we would want `P` to be
```julia
mapreduce(promote_type, 1:n) do i
    typeof(p[i, j])
end
```
assuming that all provided probabilities can be promoted to a common non-abstract type in each column (but possibly not the same type for all columns) - I guess that should usually be satisfied?

BTW copying the logic from `Distributions.DiscreteNonParametric` here unfortunately introduces also the bugs present in the implementation in Distributions (see e.g. https://github.com/JuliaStats/Distributions.jl/pull/1128), which might not be fixed when the upstream bugs are fixed if people don't keep track of Distributions closely.